Index: squashfskit-v4.14/squashfs-tools/squashfs_fs.h
===================================================================
--- squashfskit-v4.14.orig/squashfs-tools/squashfs_fs.h
+++ squashfskit-v4.14/squashfs-tools/squashfs_fs.h
@@ -299,10 +299,12 @@ struct squashfs_super_block {
 	unsigned short		block_log;
 	unsigned short		flags;
 	unsigned short		no_ids;
-	unsigned short		s_major;
+	// unsigned short		s_major;
+	long long		bytes_used;
 	unsigned short		s_minor;
 	squashfs_inode		root_inode;
-	long long		bytes_used;
+	// long long		bytes_used;
+	unsigned short		s_major;
 	long long		id_table_start;
 	long long		xattr_id_table_start;
 	long long		inode_table_start;
