Index: linux-5.4.164/drivers/usb/serial/option.c
===================================================================
--- linux-5.4.164.orig/drivers/usb/serial/option.c
+++ linux-5.4.164/drivers/usb/serial/option.c
@@ -2151,6 +2151,9 @@ static struct usb_serial_driver option_1
 #ifdef CONFIG_PM
 	.suspend           = usb_wwan_suspend,
 	.resume            = usb_wwan_resume,
+	#if 1 //Added by Quectel
+	.reset_resume = usb_wwan_resume,
+	#endif
 #endif
 };
 
Index: linux-5.4.164/drivers/usb/serial/usb_wwan.c
===================================================================
--- linux-5.4.164.orig/drivers/usb/serial/usb_wwan.c
+++ linux-5.4.164/drivers/usb/serial/usb_wwan.c
@@ -476,7 +476,13 @@ static struct urb *usb_wwan_setup_urb(st
 	usb_fill_bulk_urb(urb, serial->dev,
 			  usb_sndbulkpipe(serial->dev, endpoint) | dir,
 			  buf, len, callback, ctx);
-
+#if 1 //Added by Quectel for zero packet
+if (dir == USB_DIR_OUT) {
+struct usb_device_descriptor *desc = &serial->dev->descriptor;
+if (desc->idVendor == cpu_to_le16(0x2C7C))
+urb->transfer_flags |= URB_ZERO_PACKET;
+}
+#endif
 	if (intfdata->use_zlp && dir == USB_DIR_OUT)
 		urb->transfer_flags |= URB_ZERO_PACKET;
 
